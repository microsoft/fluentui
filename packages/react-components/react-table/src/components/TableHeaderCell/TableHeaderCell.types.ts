import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { ARIAButtonSlotProps } from '@fluentui/react-aria';
import { SortDirection, TableContextValue } from '../Table/Table.types';

export type TableHeaderCellSlots = {
  root: Slot<'th', 'div'>;

  sortIcon: Slot<'span'>;

  /**
   * Button handles correct narration and interactions for sorting;
   */
  button: NonNullable<Slot<ARIAButtonSlotProps>>;
  /**
   * aside content for anything that should be after main content of the table header cell
   */
  aside: Slot<'span'>;
};

/**
 * TableHeaderCell Props
 */
export type TableHeaderCellProps = ComponentProps<Partial<TableHeaderCellSlots>> & {
  /**
   * @default undefined
   */
  sortDirection?: SortDirection;
};

/**
 * State used in rendering TableHeaderCell
 */
export type TableHeaderCellState = ComponentState<TableHeaderCellSlots> &
  Pick<TableContextValue, 'noNativeElements' | 'sortable'>;
