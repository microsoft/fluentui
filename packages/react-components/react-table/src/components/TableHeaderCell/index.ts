export * from './TableHeaderCell';
export * from './TableHeaderCell.types';
export * from './renderTableHeaderCell';
export * from './useTableHeaderCell';
export * from './useTableHeaderCellStyles.styles';
