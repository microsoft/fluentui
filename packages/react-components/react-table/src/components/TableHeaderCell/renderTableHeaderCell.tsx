import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import type { TableHeaderCellState, TableHeaderCellSlots } from './TableHeaderCell.types';

/**
 * Render the final JSX of TableHeaderCell
 */
export const renderTableHeaderCell_unstable = (state: TableHeaderCellState) => {
  const { slots, slotProps } = getSlots<TableHeaderCellSlots>(state);

  return (
    <slots.root {...slotProps.root}>
      <slots.button {...slotProps.button}>
        {slotProps.root.children}
        {slots.sortIcon && <slots.sortIcon {...slotProps.sortIcon} />}
      </slots.button>
      {slots.aside && <slots.aside {...slotProps.aside} />}
    </slots.root>
  );
};
