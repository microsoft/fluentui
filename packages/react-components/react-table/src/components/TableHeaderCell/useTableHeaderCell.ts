import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { TableHeaderCellProps, TableHeaderCellState } from './TableHeaderCell.types';
import { useTableContext } from '../../contexts/tableContext';
import { useARIAButton } from '@fluentui/react-aria';

/**
 * Create the state required to render TableHeaderCell.
 *
 * The returned state can be modified with hooks such as useTableHeaderCellStyles_unstable,
 * before being passed to renderTableHeaderCell_unstable.
 *
 * @param props - props from this instance of TableHeaderCell
 * @param ref - reference to root HTMLElement of TableHeaderCell
 */
export const useTableHeaderCell_unstable = (
  props: TableHeaderCellProps,
  ref: React.Ref<HTMLElement>,
): TableHeaderCellState => {
  const noNativeElements = useTableContext(ctx => ctx.noNativeElements);

  const rootComponent = props.as ?? noNativeElements ? 'div' : 'th';
  return {
    components: {
      root: rootComponent,
      button: 'button',
    },
    root: getNativeElementProps(rootComponent, {
      ref,
      role: rootComponent === 'div' ? 'columnheader' : undefined,
      ...props,
    }),
    button: useARIAButton(props.button, {
      required: true,
      defaultProps: {
        // TODO implement sort state
        role: 'presentation',
        tabIndex: -1,
      },
    }),
  };
};
