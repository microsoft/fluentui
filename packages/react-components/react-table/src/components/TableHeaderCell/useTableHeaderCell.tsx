import * as React from 'react';
import { getNativeElementProps, resolveShorthand, useMergedRefs } from '@fluentui/react-utilities';
import { useFocusWithin } from '@fluentui/react-tabster';
import { ArrowUpRegular, ArrowDownRegular } from '@fluentui/react-icons';
import type { TableHeaderCellProps, TableHeaderCellState } from './TableHeaderCell.types';
import { useTableContext } from '../../contexts/tableContext';
import { useARIAButtonShorthand } from '@fluentui/react-aria';

const sortIcons = {
  ascending: <ArrowUpRegular fontSize={12} />,
  descending: <ArrowDownRegular fontSize={12} />,
};

/**
 * Create the state required to render TableHeaderCell.
 *
 * The returned state can be modified with hooks such as useTableHeaderCellStyles_unstable,
 * before being passed to renderTableHeaderCell_unstable.
 *
 * @param props - props from this instance of TableHeaderCell
 * @param ref - reference to root HTMLElement of TableHeaderCell
 */
export const useTableHeaderCell_unstable = (
  props: TableHeaderCellProps,
  ref: React.Ref<HTMLElement>,
): TableHeaderCellState => {
  const { noNativeElements, sortable } = useTableContext();

  const rootComponent = props.as ?? noNativeElements ? 'div' : 'th';
  return {
    components: {
      root: rootComponent,
      button: 'button',
      sortIcon: 'span',
      aside: 'span',
    },
    root: getNativeElementProps(rootComponent, {
      ref: useMergedRefs(ref, useFocusWithin()),
      role: rootComponent === 'div' ? 'columnheader' : undefined,
      'aria-sort': sortable ? props.sortDirection ?? 'none' : undefined,
      ...props,
    }),
    aside: resolveShorthand(props.aside),
    sortIcon: resolveShorthand(props.sortIcon, {
      required: !!props.sortDirection,
      defaultProps: { children: props.sortDirection ? sortIcons[props.sortDirection] : undefined },
    }),
    button: useARIAButtonShorthand(props.button, {
      required: true,
      defaultProps: {
        role: 'presentation',
        tabIndex: -1,
        type: 'button',
        ...(sortable && {
          role: undefined,
          tabIndex: undefined,
        }),
      },
    }),
    sortable,
    noNativeElements,
  };
};
