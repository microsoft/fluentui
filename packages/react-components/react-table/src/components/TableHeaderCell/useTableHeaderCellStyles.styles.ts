import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens } from '@fluentui/react-theme';
import type { SlotClassNames } from '@fluentui/react-utilities';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';
import type { TableHeaderCellSlots, TableHeaderCellState } from './TableHeaderCell.types';

export const tableHeaderCellClassName = 'fui-TableHeaderCell';
export const tableHeaderCellClassNames: SlotClassNames<TableHeaderCellSlots> = {
  root: 'fui-TableHeaderCell',
  button: 'fui-TableHeaderCell__button',
  sortIcon: 'fui-TableHeaderCell__sortIcon',
  aside: 'fui-TableHeaderCell__aside',
};

const useTableLayoutStyles = makeStyles({
  root: {
    display: 'table-cell',
    verticalAlign: 'middle',
  },
});

const useFlexLayoutStyles = makeStyles({
  root: {
    display: 'flex',
    ...shorthands.flex(1, 1, '0px'),
    minWidth: '0px',
  },
});

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    ...shorthands.padding('0px', tokens.spacingHorizontalS),
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius(tokens.borderRadiusMedium),
      },
      { selector: 'focus-within' },
    ),
    position: 'relative',
  },

  rootInteractive: {
    ':hover': {
      backgroundColor: tokens.colorSubtleBackgroundHover,
    },
    ':active': {
      backgroundColor: tokens.colorSubtleBackgroundPressed,
    },
  },

  resetButton: {
    resize: 'horizontal',
    boxSizing: 'content-box',
    backgroundColor: 'inherit',
    color: 'inherit',
    fontFamily: 'inherit',
    fontSize: 'inherit',
    lineHeight: 'normal',
    ...shorthands.overflow('visible'),
    ...shorthands.padding(0),
    ...shorthands.borderStyle('none'),
    WebkitAppearance: 'button',
    textAlign: 'unset',
  },

  button: {
    position: 'relative',
    width: '100%',
    display: 'flex',
    flexGrow: 1,
    height: '100%',
    alignItems: 'center',
    ...shorthands.gap(tokens.spacingHorizontalXS),
    minHeight: '32px',
    ...shorthands.flex(1, 1, '0px'),
    outlineStyle: 'none',
  },

  sortable: {
    cursor: 'pointer',
  },

  sortIcon: {
    display: 'flex',
    alignItems: 'center',
    paddingTop: tokens.spacingVerticalXXS,
  },

  resizeHandle: {},
});

/**
 * Apply styling to the TableHeaderCell slots based on the state
 */
export const useTableHeaderCellStyles_unstable = (state: TableHeaderCellState): TableHeaderCellState => {
  const styles = useStyles();
  const layoutStyles = {
    table: useTableLayoutStyles(),
    flex: useFlexLayoutStyles(),
  };
  state.root.className = mergeClasses(
    tableHeaderCellClassNames.root,
    styles.root,
    state.sortable && styles.rootInteractive,
    state.noNativeElements ? layoutStyles.flex.root : layoutStyles.table.root,
    state.root.className,
  );
  state.button.className = mergeClasses(
    tableHeaderCellClassNames.button,
    styles.resetButton,
    styles.button,
    state.sortable && styles.sortable,
    state.button.className,
  );

  if (state.sortIcon) {
    state.sortIcon.className = mergeClasses(
      tableHeaderCellClassNames.sortIcon,
      styles.sortIcon,
      state.sortIcon.className,
    );
  }

  if (state.aside) {
    state.aside.className = mergeClasses(tableHeaderCellClassNames.aside, styles.resizeHandle, state.aside.className);
  }

  return state;
};
