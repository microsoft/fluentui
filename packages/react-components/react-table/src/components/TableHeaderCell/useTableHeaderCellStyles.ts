import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens } from '@fluentui/react-theme';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { TableHeaderCellSlots, TableHeaderCellState } from './TableHeaderCell.types';

export const tableHeaderCellClassName = 'fui-TableHeaderCell';
export const tableHeaderCellClassNames: SlotClassNames<TableHeaderCellSlots> = {
  root: 'fui-TableHeaderCell',
  button: 'fui-TableHeaderCell__button',
  sortIcon: 'fui-TableHeaderCell__sortIcon',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'table-cell',
    verticalAlign: 'middle',
    ...shorthands.padding('0px', tokens.spacingHorizontalS),
  },

  resetButton: {
    resize: 'horizontal',
    boxSizing: 'content-box',
    backgroundColor: 'inherit',
    color: 'inherit',
    fontFamily: 'inherit',
    fontSize: 'inherit',
    lineHeight: 'normal',
    ...shorthands.overflow('visible'),
    ...shorthands.padding(0),
    ...shorthands.borderStyle('none'),
    WebkitAppearance: 'button',
    textAlign: 'unset',
  },
  button: {
    width: '100%',
    display: 'flex',
    flexGrow: 1,
    height: '100%',
    alignItems: 'center',
    ...shorthands.gap(tokens.spacingHorizontalS),
    minHeight: '44px',
    ...shorthands.flex(1, 1, '0px'),
  },
  sortable: {
    cursor: 'pointer',
  },

  sortIcon: {
    display: 'flex',
    alignItems: 'center',
  },
});

/**
 * Apply styling to the TableHeaderCell slots based on the state
 */
export const useTableHeaderCellStyles_unstable = (state: TableHeaderCellState): TableHeaderCellState => {
  const styles = useStyles();
  state.root.className = mergeClasses(tableHeaderCellClassNames.root, styles.root, state.root.className);
  state.button.className = mergeClasses(
    tableHeaderCellClassNames.button,
    styles.resetButton,
    styles.button,
    state.sortable && styles.sortable,
    state.button.className,
  );

  if (state.sortIcon) {
    state.sortIcon.className = mergeClasses(
      tableHeaderCellClassNames.sortIcon,
      styles.sortIcon,
      state.sortIcon.className,
    );
  }

  return state;
};
