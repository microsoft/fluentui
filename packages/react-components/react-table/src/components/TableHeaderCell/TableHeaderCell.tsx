import * as React from 'react';
import { useTableHeaderCell_unstable } from './useTableHeaderCell';
import { renderTableHeaderCell_unstable } from './renderTableHeaderCell';
import { useTableHeaderCellStyles_unstable } from './useTableHeaderCellStyles';
import type { TableHeaderCellProps } from './TableHeaderCell.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * TableHeaderCell component
 */
export const TableHeaderCell: ForwardRefComponent<TableHeaderCellProps> = React.forwardRef((props, ref) => {
  const state = useTableHeaderCell_unstable(props, ref);

  useTableHeaderCellStyles_unstable(state);
  return renderTableHeaderCell_unstable(state);
});

TableHeaderCell.displayName = 'TableHeaderCell';
