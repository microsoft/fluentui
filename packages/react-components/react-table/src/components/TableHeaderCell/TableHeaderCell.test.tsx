import * as React from 'react';
import { render } from '@testing-library/react';
import { TableHeaderCell } from './TableHeaderCell';
import { isConformant } from '../../testing/isConformant';
import { TableHeaderCellProps } from './TableHeaderCell.types';
import { TableContextProvider, tableContextDefaultValue } from '../../contexts/tableContext';
import { SortDirection } from '../Table/Table.types';

describe('TableHeaderCell', () => {
  const tr = document.createElement('tr');
  beforeEach(() => {
    document.body.appendChild(tr);
  });

  isConformant({
    Component: TableHeaderCell as React.FC<TableHeaderCellProps>,
    displayName: 'TableHeaderCell',
    testOptions: {
      'has-static-classnames': [
        {
          props: {
            sortIcon: 'Test Icon',
            aside: 'Test aside',
          },
        },
      ],
    },
    renderOptions: {
      container: tr,
    },
  });

  it('renders a default state', () => {
    const result = render(<TableHeaderCell>Default TableHeaderCell</TableHeaderCell>, {
      container: tr,
    });
    expect(result.container).toMatchSnapshot();
  });

  it('renders as div if `noNativeElements` is set', () => {
    const { container } = render(
      <TableContextProvider value={{ ...tableContextDefaultValue, noNativeElements: true }}>
        <TableHeaderCell>Cell</TableHeaderCell>
      </TableContextProvider>,
    );
    expect(container.firstElementChild?.tagName).toEqual('DIV');
    expect(container.firstElementChild?.getAttribute('role')).toEqual('columnheader');
  });

  it('should render sortIcon when sortable is true and the header cell is sorted', () => {
    const { container } = render(
      <TableContextProvider value={{ ...tableContextDefaultValue, noNativeElements: true, sortable: true }}>
        <TableHeaderCell sortDirection="ascending">Cell</TableHeaderCell>
      </TableContextProvider>,
    );
    expect(container.querySelector('svg')).not.toBe(null);
  });

  it('should not render sortIcon when the header cell is not sorted', () => {
    const { container } = render(
      <TableContextProvider value={{ ...tableContextDefaultValue, noNativeElements: true, sortable: true }}>
        <TableHeaderCell>Cell</TableHeaderCell>
      </TableContextProvider>,
    );
    expect(container.querySelector('svg')).toBe(null);
  });

  it('should not have interactive button when not sortable', () => {
    const { container } = render(
      <TableContextProvider value={{ ...tableContextDefaultValue, noNativeElements: true, sortable: false }}>
        <TableHeaderCell>Cell</TableHeaderCell>
      </TableContextProvider>,
    );

    const button = container.querySelector('button');
    expect(button?.getAttribute('role')).toEqual('presentation');
    expect(button?.getAttribute('tabindex')).toEqual('-1');
  });

  it.each<SortDirection>(['ascending', 'descending'])(
    'should render aria-sort according to sortDirection',
    sortDirection => {
      const { getByRole } = render(
        <TableContextProvider
          value={{
            ...tableContextDefaultValue,
            noNativeElements: true,
            sortable: true,
          }}
        >
          <TableHeaderCell sortDirection={sortDirection}>Cell</TableHeaderCell>
        </TableContextProvider>,
      );
      expect(getByRole('columnheader').getAttribute('aria-sort')).toEqual(sortDirection);
    },
  );

  it('should render aria-sort "none" when column is not sorted', () => {
    const { getByRole } = render(
      <TableContextProvider
        value={{
          ...tableContextDefaultValue,
          noNativeElements: true,
          sortable: true,
        }}
      >
        <TableHeaderCell>Cell</TableHeaderCell>
      </TableContextProvider>,
    );
    expect(getByRole('columnheader').getAttribute('aria-sort')).toEqual('none');
  });

  it('should not render aria-sort table is not sortable', () => {
    const { getByRole } = render(
      <TableContextProvider
        value={{
          ...tableContextDefaultValue,
          noNativeElements: true,
          sortable: false,
        }}
      >
        <TableHeaderCell>Cell</TableHeaderCell>
      </TableContextProvider>,
    );
    expect(getByRole('columnheader').hasAttribute('aria-sort')).toBe(false);
  });
});
