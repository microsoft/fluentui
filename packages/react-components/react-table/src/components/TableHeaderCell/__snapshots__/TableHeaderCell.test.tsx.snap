// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeaderCell renders a default state 1`] = `
<tr>
  <th
    class="fui-TableHeaderCell"
  >
    <button
      class="fui-TableHeaderCell__button"
      role="presentation"
      tabindex="-1"
      type="button"
    >
      Default TableHeaderCell
    </button>
  </th>
</tr>
`;
